kolibri/core/assets/src/exams/utils.js

```diff
@@ -84,106 +81,114 @@ function annotateQuestionsWithItem(questions) {
-  const questions = annotateQuestionsWithItem(questionSources);
-  return {
-    section_id: uuidv4(),
-    section_title: '',
-    description: '',
-    resource_pool: [],
-    questions,
-    learners_see_fixed_order: exam.learners_see_fixed_order,
-    question_count: exam.question_count,
-  };
-}
-
-export function revertV3toV2(questionSources) {
-  if (!questionSources.length) {
-    return [];
-  }
-  return questionSources[0].questions;
+export function convertExamQuestionSourcesV2toV3({ question_sources, learners_see_fixed_order }) {
+  // In V2, question_sources are questions so we add them
+  // to the newly created section's `questions` property
+  const questions = question_sources;
+  return [
+    {
+      section_id: uuidv4(),
+      section_title: '',
+      description: '',
+      questions,
+      learners_see_fixed_order,
+      question_count: questions.length,
+    },
+  ];
 }
 
 /**
- * @param {object} exam - an exam object of any question_sources version
- * @returns V3 formatted question_sources
+ * Fetches the content nodes for an exam and converts the exam to the latest data_model_version
+ *
+ * data_model_version 0 (V0):
+ * - question_sources here refer to exercise nodes that the exam drew questions from at that time
+ *
+ * data_model_version 1 (V1):
+ * - question_sources is changed to now refer to the questions themselves by including the
+ *   exercise_id and question_id along with a title
+ *
+ * data_model_version 2 (V2):
+ * - The objects in question_sources are now annotated with a counter_in_exercise field
+ *
+ * data_model_version 3 (V3):
+ * - question_sources now refers to a list of sections, each with their own list of questions
  */
-export function convertExamQuestionSourcesToV3(exam, extraArgs = {}) {
-  if (exam.data_model_version !== 3) {
-    const V2_sources = convertExamQuestionSources(exam, extraArgs);
-    return [convertV2toV3(V2_sources, exam)];
-  }
 
-  return exam.question_sources;
+export async function convertExamQuestionSources(exam) {
+  const { data_model_version } = exam;
+
+  return new Promise(resolve => {
+    if (data_model_version === 0) {
+      const ids = uniq(exam.question_sources.map(item => item.exercise_id));
+      return ContentNodeResource.fetchCollection({
+        getParams: {
+          ids,
+          no_available_filtering: true,
+        },
+      }).then(exercises => {
+        const questionIds = exercises.reduce((nodeIds, node) => {
+          nodeIds[node.id] = node.assessmentmetadata
+            ? node.assessmentmetadata.assessment_item_ids
+            : [];
+          return nodeIds;
+        }, []);
+        exam.question_sources = convertExamQuestionSourcesV0V1(
+          exam.question_sources,
+          exam.seed,
+          questionIds
+        );
+        resolve(exam);
+      });
+    } else {
+      resolve(exam);
+    }
+  }).then(exam => {
+    if (data_model_version <= 1) {
```

Comment:
Surely after each step here, we would need to update the data_model_version, otherwise we will update to v2 and no further, if the data_model_version is 1?

Reply:
The subsequent if-statements will run too because 1 <= 2 == true then the next because 1 <= 3, etc. 

If we updated the data_model_version I don't think that would impact how this works, though, and the checks could be made to be `===` checks rather than relying on this fall-through approach.

I think that my reasoning for not updating the `data_model_version` was thinking the data on the front-end should reflect what is persisted to the DB in this case -- but thinking a bit more now it does make sense to have that particular value reflect the structure of the data as it is now as it is passed around the front-end.

Reply:
Yeah - I think my question here was connected to the v0v2 function - which converts from v0v2 (not v0 to v1, I think) and so as you correctly point out it got passed along here, we would be passing v2 stuff into the v1 to v2 because it had already been converted (although now, I do understand your desire to make each function do a single version leap).